text/0000-js-compatible-resolution.md

```diff
@@ -0,0 +1,344 @@
+
+3. When both `.js` and `.hbs` exist for a component, the build system unobtrusively associates the template with the default value exported by the Javascript. The template must be co-located with the Javascript, having the same filename except for changing the extension to `.hbs`.
+
+    For illustrative purposes, here's a toy implementation of how we can do "unobtrusive" association of the template with the Javascript value:
+
+    ```js
+    // the previous example code would get rewritten to something like this:
+    import { registerTemplate } from '@ember/hypothetical-private-api';
+    export default registerTemplate(class extends Component {
+      yourStuffHere()
+    }, compiledTemplateAndMetadataHere);
+    ```
+
+    where the implementation of `registerTemplate` effectively does:
+
+    ```js
+    const templates = new WeakMap();
+    function registerTemplate(value, templateDetails) {
+      templates.set(value, templateDetails);
+      // notice that the value is unchanged. That's why I mean by "unobtrustive".
+      // You can still import this value into both templates and javascript and
+      // get what you expected to get.
+      return value;
+    }
+    ```
+
+    Please note this is only an example to show that unobtrusive association is possible. The exact details would be private API.
+
+4. When only a template exists, it gets resolved following the normal Module Resolver system as described in step 1. Our handler for `.hbs` files causes them to be interpreted as Javasript modules like:
+
+    ```js
+    // for illustrative purposes only
+    import { registerTemplate, useDefaultComponentManager } from '@ember/hypothetical-private-api';
+    export default registerTemplate(useDefaultComponentManager, compiledTemplateAndMetadataHere);
+    ```
+
+    The point of this is that you can import and invoke the template-only component in the same way you would a template-and-javascript component. In both cases, from glimmer's perspective the actual value that you import is used to lookup the component's metadata & template in a shared, behind-the-scenes WeakMap.
+
+5. Helpers are resolved just like component Javascript. They are simpler because there is no template association to worry about.
+
+6. Anything that was not resolved by this new system continues to be resolved the classic way.
+
+## Implications
+
+Given the system described above, several patterns _necessarily_ become possible. Specifically, this component invocation:
+
+```
+---
+import Button from './path/to/button';
+---
+<Button />
+```
+
+Should work correctly for all of the following situations:
+
+1. `./path/to/button.hbs` is a template-only component.
+2. `./path/to/button/index.hbs` is a template-only component.
+3. `./path/to/button.js` exports a component class and `./path/to/button.hbs` is its template.
+4. `./path/to/button/index.js` exports a component class and `./path/to/button/index.hbs` is its template.
```

Comment:
In addition to https://github.com/ef4/rfcs/pull/16/files#r246491839, does this also mean that private local lookup will be deprecated, since any component can essentially import any other component?